### PR TITLE
feat(sql-parser): Replacing sqlmetadata sql parser lib with sqlineage parser lib

### PR DESCRIPTION
**Pull request by treff7es**
## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: metadata-ingestion/src/datahub/utilities/sql_parser.py [hunk 1]
```diff
@@ -72,3 +83,116 @@ def get_columns(self) -> List[str]:
...
+                sqlparse.format(
+                    self._sql.strip(),
+                    strip_comments=True,
+                    use_space_around_operators=True,
+                ),
+            )
+            if s.token_first(skip_cm=True)
+        ]
+
+        with unittest.mock.patch(
```
Comment (RickardCardell) [inline metadata-ingestion/src/datahub/utilities/sql_parser.py:138]: Couldn't this be done without using `unittest` and mocking?

Comment (treff7es) [reply metadata-ingestion/src/datahub/utilities/sql_parser.py:138]: Currently Sqllineage can't get tables/columns for select queries. 
This is only needed until we contribute back this and sqllineage can get source columns for select queries.

Comment (RickardCardell) [reply metadata-ingestion/src/datahub/utilities/sql_parser.py:138]: This change prevents some companies from using this library.

Imho, better to leave out the column lineage than to use mocking and testing libraries in production code. Or even wait until the related fix is implemented in sqllineage.

**Comment by shirshanka** [comment]
@rslanka: can you take a look at this one?

File: metadata-ingestion/src/datahub/utilities/sql_parser.py [hunk 2]
```diff
@@ -48,10 +57,12 @@ def __init__(self, sql_query: str) -> None:
         self._parser = MetadataSQLParser(sql_query)
 
     def get_tables(self) -> List[str]:
-        return self._parser.tables
+        result = self._parser.tables
+        # Sort tables to make the list deterministic
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/utilities/sql_parser.py:61]: Why is deterministic table ordering required here?

File: metadata-ingestion/src/datahub/utilities/sql_parser.py [hunk 3]
```diff
@@ -72,3 +83,116 @@ def get_columns(self) -> List[str]:
...
+            ):
+                self._stmt_holders = [
+                    LineageAnalyzer().analyze(stmt) for stmt in self._stmt
+                ]
+                self._sql_holder = SQLLineageHolder.of(*self._stmt_holders)
+
+    def get_tables(self) -> List[str]:
+        result: List[str] = list()
+        for table in self._sql_holder.source_tables:
+            table_normalized = re.sub(r"^<default>.", "", str(table))
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/utilities/sql_parser.py:154]: Could you add a comment on why this kind of normalization is required?

File: metadata-ingestion/src/datahub/utilities/sql_parser.py [hunk 4]
```diff
@@ -72,3 +83,116 @@ def get_columns(self) -> List[str]:
...
+        target_columns = {column for column, deg in column_graph.out_degree if deg == 0}
+
+        result: Set[str] = set()
+        for column in target_columns:
+            # Let's drop all the count(*) and similard columns which are expression actually if it does not have an alias
+            if not any(ele in column.raw_name for ele in ["*", "(", ")"]):
+                result.add(str(column.raw_name))
+
+        # Reverting back all the previously renamed words which confuses the parser
+        result = set(["date" if c == self._DATE_SWAP_TOKEN else c for c in result])
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/utilities/sql_parser.py:176]: There is one iteration over the entire result set for replacing each swapped token. Could you replace lines 176-182 with a single statement like
 ```python 
result = set([unswap_tokens(c) for c in result])
```
where `unswap_tokens` handles the token conversion as needed?

File: metadata-ingestion/src/datahub/utilities/sql_parser.py [hunk 5]
```diff
@@ -72,3 +83,116 @@ def get_columns(self) -> List[str]:
...
+                ]
+                self._sql_holder = SQLLineageHolder.of(*self._stmt_holders)
+
+    def get_tables(self) -> List[str]:
+        result: List[str] = list()
+        for table in self._sql_holder.source_tables:
+            table_normalized = re.sub(r"^<default>.", "", str(table))
+            result.append(str(table_normalized))
+
+        # Sort tables to make the list deterministic
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/utilities/sql_parser.py:157]: Why is deterministic order required here?

File: metadata-ingestion/src/datahub/utilities/sqllineage_patch.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+from sqllineage.core.holders import SubQueryLineageHolder
+from sqllineage.core.models import Column
+from sqllineage.exceptions import SQLLineageException
+from sqllineage.utils.constant import EdgeType
+
+
+# Patch based on sqllineage v1.3.3
+def end_of_query_cleanup_patch(self, holder: SubQueryLineageHolder) -> None:  # type: ignore
+    for i, tbl in enumerate(self.tables):
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/utilities/sqllineage_patch.py:11]: Why `enumerate` here, the index is not used?

File: metadata-ingestion/src/datahub/utilities/sqllineage_patch.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+    for i, (col_barrier, tbl_barrier) in enumerate(self.union_barriers):
+        prev_col_barrier, prev_tbl_barrier = (
+            (0, 0) if i == 0 else self.union_barriers[i - 1]
+        )
+        col_grp = self.columns[prev_col_barrier:col_barrier]
+        tbl_grp = self.tables[prev_tbl_barrier:tbl_barrier]
+        tgt_tbl = None
+        if holder.write:
+            if len(holder.write) > 1:
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/utilities/sqllineage_patch.py:23]: Could you add a comment on what this check is signifying?

File: metadata-ingestion/src/datahub/utilities/sqllineage_patch.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+                raise SQLLineageException
+            tgt_tbl = list(holder.write)[0]
+        elif holder.read:
+            tgt_tbl = list(holder.read)[0]
+
+        if tgt_tbl:
+            for tgt_col in col_grp:
+                tgt_col.parent = tgt_tbl
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/utilities/sqllineage_patch.py:31]: some comments on what is going on here would be really useful.

File: metadata-ingestion/src/datahub/utilities/sqllineage_patch.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        )
+        col_grp = self.columns[prev_col_barrier:col_barrier]
+        tbl_grp = self.tables[prev_tbl_barrier:tbl_barrier]
+        tgt_tbl = None
+        if holder.write:
+            if len(holder.write) > 1:
+                raise SQLLineageException
+            tgt_tbl = list(holder.write)[0]
+        elif holder.read:
+            tgt_tbl = list(holder.read)[0]
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/utilities/sqllineage_patch.py:27]: For all `i`, is the `tgt_tbl` always expected to be the same? If now, how are the header.read/write getting updated, since we always seem to take the value at index 0.

File: metadata-ingestion/src/datahub/utilities/sqllineage_patch.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+                    if holder.write:
+                        holder.add_column_lineage(src_col, tgt_col)
+                    else:
+                        holder.add_column_lineage(None, tgt_col)
+
+
+def add_column_lineage_patch(self, src: Optional[Column], tgt: Column) -> None:  # type: ignore
+    if tgt and src:
+        self.graph.add_edge(src, tgt, type=EdgeType.LINEAGE)
+    if tgt:
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/utilities/sqllineage_patch.py:44]: Looks like `tgt` also is Optional.

File: metadata-ingestion/tests/unit/test_redash_source.py [hunk 11]
```diff
@@ -685,8 +685,8 @@ def test_get_chart_snapshot_parse_table_names_from_sql(mocked_data_source):
                 ),
                 chartUrl="http://localhost:5000/queries/4#10",
                 inputs=[
-                    "urn:li:dataset:(urn:li:dataPlatform:mysql,Rfam.orders,PROD)",
                     "urn:li:dataset:(urn:li:dataPlatform:mysql,Rfam.order_items,PROD)",
+                    "urn:li:dataset:(urn:li:dataPlatform:mysql,Rfam.orders,PROD)",
```
Comment (rslanka) [inline metadata-ingestion/tests/unit/test_redash_source.py:689]: Why is this necessary? Would this sort order requirement break any of the existing customers doing custom transforms?

File: metadata-ingestion/tests/unit/test_utilities.py [hunk 12]
```diff
@@ -55,9 +55,297 @@ def test_groupby_unsorted():
...
+    columns_list.sort()
+    assert columns_list == ["a", "b", "c"]
+
+
+def test_sqllineage_sql_parser_get_columns_from_simple_query():
+    sql_query = "SELECT foo.a, foo.b FROM foo;"
+
+    columns_list = SqlLineageSQLParser(sql_query).get_columns()
+    columns_list.sort()
+    assert columns_list == ["a", "b"]
```
Comment (rslanka) [inline metadata-ingestion/tests/unit/test_utilities.py:152]: Looks like there are only two kinds of tests, the ones testing get_tables and the ones testing get_columns, each type doing the same set of operations for different params. Consider refactoring with the `pytest.mark.parametrize` decorator - [examples](https://docs.pytest.org/en/6.2.x/example/parametrize.html#paramexamples)

File: metadata-ingestion/src/datahub/utilities/sql_parser.py (hunk 2 above)
Comment (treff7es) [reply metadata-ingestion/src/datahub/utilities/sql_parser.py:61]: There were test in metabase which expected to have the same order always. I wanted to make sure all the test are fine without any modification to make sure I did not break anything.

File: metadata-ingestion/tests/unit/test_utilities.py (hunk 12 above)
Comment (treff7es) [reply metadata-ingestion/tests/unit/test_utilities.py:152]: sure, will do, thanks.

**Review by shirshanka — APPROVED** [review]
LGTM. @treff7es you can address the style comments in a future PR.
